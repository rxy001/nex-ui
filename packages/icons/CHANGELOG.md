# @nex-ui/icons

## 0.0.6

### Patch Changes

- 87191c1: release
- Updated dependencies [87191c1]
  - @nex-ui/styled@0.0.6
  - @nex-ui/utils@0.0.6

## 0.0.5

### Patch Changes

- 7631577: release
- Updated dependencies [7631577]
  - @nex-ui/styled@0.0.5
  - @nex-ui/utils@0.0.5

## 0.0.4

### Patch Changes

- 70cb478: v0.0.4
- Updated dependencies [70cb478]
  - @nex-ui/styled@0.0.4
  - @nex-ui/utils@0.0.4
