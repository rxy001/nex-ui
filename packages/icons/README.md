# @nex-ui/icons

Shared icons used by NexUI packages.

## Installation

```bash
npm i @nex-ui/icons
```
