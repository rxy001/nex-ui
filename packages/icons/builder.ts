/* eslint-disable no-console */
import {
  readdirSync,
  statSync,
  writeFileSync,
  existsSync,
  readFileSync,
  mkdirSync,
} from 'node:fs'
import { fileURLToPath } from 'node:url'
import { forEach, map } from '@nex-ui/utils'
import { dirname, resolve, basename } from 'node:path'
import type { IconProps } from './src/utils'

const options: Record<string, IconProps> = {
  LoadingOutlined: {
    spin: true,
  },
  Loading3QuartersOutlined: {
    spin: true,
  },
}

function upperFirst(string: string) {
  return string.charAt(0).toUpperCase() + string.slice(1)
}

function run() {
  const cwd = dirname(fileURLToPath(import.meta.url))

  const svgDirPath = resolve(cwd, './src/svg')

  const componentsDirPath = resolve(cwd, './src/components')

  const newTsxs: Record<string, string[]> = {}

  console.log(`[builder] Building...`)

  recursive(svgDirPath, generateTsx)

  writeImports()

  console.log(`[builder] Generated files âœ…`)

  function recursive(
    dirPath: string,
    cb: (filePath: string, parentDir: string) => void,
  ) {
    const files = readdirSync(dirPath)
    const parentDir = basename(dirPath)

    forEach(files, (file: string) => {
      const filePath = resolve(dirPath, file)
      if (statSync(filePath).isDirectory()) {
        recursive(filePath, cb)
      } else {
        cb(filePath, parentDir)
      }
    })
  }

  function generateTsx(svgFilePath: string, category: string) {
    const svgFileName = basename(svgFilePath, '.svg')

    const svgComponentName = svgFileName
      .split('-')
      .map((v) => upperFirst(v))
      .join('')

    const iconComponentName = `${svgComponentName}${upperFirst(category)}`

    const tsxName = `${iconComponentName}.tsx`

    const categoryDirPath = resolve(componentsDirPath, category)

    if (!existsSync(categoryDirPath)) {
      mkdirSync(categoryDirPath)
    }

    const tsxPath = resolve(categoryDirPath, tsxName)

    if (!existsSync(tsxPath)) {
      const tsx =
        `import { createIcon } from '../../utils'` +
        '\n' +
        `import ${svgComponentName} from '../../svg/${category}/${svgFileName}.svg'` +
        '\n\n' +
        `export const ${iconComponentName} = createIcon(${svgComponentName}${
          options[iconComponentName]
            ? `,${JSON.stringify(options[iconComponentName])}`
            : ''
        })` +
        '\n'

      writeFileSync(tsxPath, tsx)

      if (newTsxs[category]) {
        newTsxs[category].push(iconComponentName)
      } else {
        newTsxs[category] = [iconComponentName]
      }
    }
  }

  function writeImports() {
    const indexPath = resolve(componentsDirPath, './index.tsx')

    let tsx = ''
    if (existsSync(indexPath)) {
      tsx = readFileSync(indexPath, {
        encoding: 'utf-8',
      })
    }

    forEach(newTsxs, (tsxs: string[], category: string) => {
      tsx =
        // eslint-disable-next-line prefer-template
        tsx +
        '\n' +
        map(
          tsxs,
          (name: string) => `export * from './${category}/${name}'`,
        ).join('\n')
    })

    writeFileSync(indexPath, tsx)
  }
}

run()
