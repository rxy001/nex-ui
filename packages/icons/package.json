{
  "name": "@nex-ui/icons",
  "version": "1.0.0",
  "author": "X1ng Yu",
  "private": false,
  "keywords": [
    "react",
    "react-component",
    "nui",
    "nex-ui",
    "icons"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/rxy001/nex-ui.git",
    "directory": "packages/icons"
  },
  "license": "MIT",
  "scripts": {
    "build": "tsx ../../scripts/build.ts --dts",
    "build:fast": "tsx ../../scripts/build.ts",
    "type-check": "tsc --noEmit"
  },
  "type": "module",
  "main": "dist/cjs/index.cjs",
  "module": "dist/esm/index.mjs",
  "types": "dist/types/index.d.ts",
  "exports": {
    ".": {
      "source": "./src/index.ts",
      "import": {
        "types": "./dist/types/index.d.ts",
        "default": "./dist/esm/index.mjs"
      },
      "require": {
        "types": "./dist/types/index.d.ts",
        "default": "./dist/cjs/index.cjs"
      }
    },
    "./package.json": "./package.json"
  },
  "peerDependencies": {
    "@emotion/react": ">=11",
    "react": ">=18",
    "react-dom": ">=18"
  },
  "sideEffects": false,
  "devDependencies": {
    "@storybook/react": "8.0.10"
  },
  "dependencies": {
    "@nex-ui/utils": "workspace:*"
  }
}
