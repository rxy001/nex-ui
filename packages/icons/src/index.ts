export * from './components'
export { NexIconsProvider as __NEX_ICON_PROVIDER } from './utils'

export type { NexIconsProviderProps as __NexIconsProviderProps } from './utils'

export type { IconProps, IconPropsOverrides } from './types'
