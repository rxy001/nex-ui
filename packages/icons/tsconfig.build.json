{
  "extends": "../../tsconfig.json",
  "include": ["./src", "./types.d.ts"],
  "exclude": ["./src/**/__tests__/**/*"]
}
