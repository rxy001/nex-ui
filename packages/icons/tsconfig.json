{
  "extends": "../../tsconfig.json",
  "include": ["./src", "./types.d.ts"],
  "compilerOptions": {
    "customConditions": ["source"],
    "jsxImportSource": "@emotion/react",
    "jsx": "react-jsx"
  }
}
