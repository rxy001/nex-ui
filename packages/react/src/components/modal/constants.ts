export const MODAL_INTERNAL_ID_PREFIX = 'nui-modal-'
