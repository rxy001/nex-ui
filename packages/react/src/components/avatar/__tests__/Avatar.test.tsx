import {
  testComponentStability,
  renderWithNexUIProvider,
  mockGlobalImage,
  restoreGlobalImage,
  testRootClassName,
  flushMockedImages,
  testColorDataAttrs,
  testSizeDataAttrs,
  testVariantDataAttrs,
  testRefForwarding,
} from '~/tests/shared'
import { UserOutlined } from '@nex-ui/icons'
import { Avatar } from '../index'
import { avatarSlotClasses, avatarDataAttrs } from './constants'

afterAll(() => {
  restoreGlobalImage()
})

describe('Avatar', () => {
  testComponentStability(<Avatar />)

  testRootClassName(<Avatar />)

  testRefForwarding(<Avatar />)

  testColorDataAttrs(<Avatar>Avatar</Avatar>)

  testSizeDataAttrs(<Avatar>Avatar</Avatar>)

  testVariantDataAttrs(<Avatar>Avatar</Avatar>, [
    'radius',
    ['none', 'sm', 'md', 'lg', 'xl', 'full'],
  ])

  testVariantDataAttrs(<Avatar outlined>Avatar</Avatar>, [
    'outlined',
    [true, false],
  ])

  it('should render with default props', () => {
    const { container } = renderWithNexUIProvider(<Avatar>A</Avatar>)

    const avatarRoot = container.firstElementChild
    expect(avatarRoot).toHaveClass(avatarSlotClasses.root)
    expect(avatarRoot).toHaveAttribute(...avatarDataAttrs['inGroup-false'])
    expect(avatarRoot).toHaveAttribute(...avatarDataAttrs['size-md'])
    expect(avatarRoot).toHaveAttribute(...avatarDataAttrs['radius-md'])
    expect(avatarRoot).toHaveAttribute(...avatarDataAttrs['color-gray'])
    expect(avatarRoot).toHaveAttribute(...avatarDataAttrs['outlined-false'])

    expect(avatarRoot).toMatchSnapshot()
  })

  describe('Image Avatar', () => {
    afterEach(() => {
      restoreGlobalImage()
    })

    it('should forward classNames to img slot', () => {
      mockGlobalImage('loaded')
      const classNames = {
        img: 'test-img-class',
      }
      const { container } = renderWithNexUIProvider(
        <Avatar
          src='/fake.png'
          classNames={{
            img: classNames.img,
          }}
        />,
      )
      const avatarRoot = container.firstElementChild
      expect(avatarRoot?.firstElementChild).toHaveClass(classNames.img)
      restoreGlobalImage()
    })

    it('should forward slotProps to img slot', () => {
      mockGlobalImage('loaded')

      const slotProps = {
        img: {
          className: 'test-img-class',
        },
      }

      const { queryByClassName } = renderWithNexUIProvider(
        <Avatar data-testid='avatar' src='/fake.png' slotProps={slotProps}>
          Avatar
        </Avatar>,
      )
      expect(queryByClassName(avatarSlotClasses.img)).toHaveClass(
        slotProps.img.className,
      )
    })

    it('should render an img when the image loads successfully', () => {
      mockGlobalImage('loaded')
      const { container, rerender } = renderWithNexUIProvider(
        <Avatar src='/fake.png' alt='Avatar' />,
      )
      const avatarRoot = container.firstElementChild
      const avatarImg = avatarRoot?.firstElementChild
      expect(avatarRoot?.tagName).toBe('DIV')
      expect(avatarImg?.tagName).toBe('IMG')
      expect(avatarImg).toHaveAttribute('alt', 'Avatar')
      expect(avatarImg).toHaveAttribute('src', '/fake.png')

      rerender(<Avatar srcSet='/fake.png 1x, /fake@2x.png 2x' />)

      expect(avatarImg).toHaveAttribute(
        'srcSet',
        '/fake.png 1x, /fake@2x.png 2x',
      )
    })

    it('should render its children when the image fails to load', () => {
      mockGlobalImage('error')
      const { container } = renderWithNexUIProvider(
        <Avatar src='/fake.png' alt='Avatar' data-testid='avatar-children'>
          B
        </Avatar>,
      )
      expect(container.firstElementChild?.textContent).toBe('B')
    })

    it('should render the first letter of its alt when the image fails to load', () => {
      mockGlobalImage('error')
      const { container } = renderWithNexUIProvider(
        <Avatar src='/fake.png' alt='Avatar' data-testid='avatar-text' />,
      )
      expect(container.firstElementChild?.textContent).toBe('A')
    })

    it('should not update state if image is unmounted before load event', () => {
      mockGlobalImage('loading')

      const { unmount, container } = renderWithNexUIProvider(
        <Avatar src='/fake.png' alt='Avatar' />,
      )

      expect(container.firstElementChild?.textContent).toBe('A')

      unmount()

      flushMockedImages('loaded')

      expect(true).toBe(true)
    })

    it('should not update state if image is unmounted before error event', () => {
      mockGlobalImage('loading')

      const { unmount, container } = renderWithNexUIProvider(
        <Avatar src='/fake.png' alt='Avatar' />,
      )

      expect(container.firstElementChild?.textContent).toBe('A')

      unmount()

      flushMockedImages('error')

      expect(true).toBe(true)
    })
  })

  describe('Text Avatar', () => {
    it('should render a div containing a string', () => {
      const { container } = renderWithNexUIProvider(<Avatar>O</Avatar>)
      const avatarRoot = container.firstElementChild
      expect(avatarRoot?.tagName).toBe('DIV')
      expect(avatarRoot?.firstChild?.textContent).toBe('O')
    })
  })

  describe('Icon Avatar', () => {
    it('should render a div containing an svg icon', () => {
      const { container } = renderWithNexUIProvider(
        <Avatar>
          <UserOutlined className='nui-icon' />
        </Avatar>,
      )
      const avatarRoot = container.firstElementChild
      expect(avatarRoot?.tagName).toBe('DIV')
      const userIcon = avatarRoot?.firstElementChild
      expect(userIcon).toHaveClass('nui-icon')
    })
  })
})
