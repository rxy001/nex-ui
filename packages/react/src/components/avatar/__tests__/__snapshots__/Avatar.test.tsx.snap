// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render with default props 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--nui-colors-gray-contrast-text);
  background-color: var(--nui-colors-gray-primary);
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: var(--nui-sizes-10);
  height: var(--nui-sizes-10);
  font-size: var(--nui-font-sizes-lg);
  border-radius: var(--nui-radii-lg);
}

.emotion-0 svg {
  font-size: 1.6em;
}

<div
  aria-label="A"
  class="nui-avatar-root emotion-0"
  data-color="gray"
  data-in-group="false"
  data-outlined="false"
  data-radius="md"
  data-size="md"
  role="img"
>
  A
</div>
`;
